zk/stages/stage_witness.go

```diff
@@ -27,19 +27,228 @@ import (
+	return WitnessCfg{
+		db:          db,
+		tmpDir:      tmpDir,
+		blockReader: blockReader,
+		engine:      engine,
+		chainConfig: chainConfig,
+		agg:         agg,
+		historyV3:   historyV3,
+	}
+}
+
+func SpawnWitnessStage(
+	s *stagedsync.StageState,
+	u stagedsync.Unwinder,
+	tx kv.RwTx,
+	ctx context.Context,
+	cfg WitnessCfg,
+	initialCycle bool,
+	quiet bool,
+) error {
+	logPrefix := s.LogPrefix()
+	log.Info(fmt.Sprintf("[%s] Starting witness stage", logPrefix))
+	defer log.Info(fmt.Sprintf("[%s] Finished witness stage", logPrefix))
+
+	if sequencer.IsSequencer() {
+		log.Info(fmt.Sprintf("[%s] skipping -- sequencer", logPrefix))
+		return nil
+	}
+
+	if tx == nil {
+		log.Debug(fmt.Sprintf("[%s] witness: no tx provided, creating a new one", logPrefix))
+		var err error
+		tx, err = cfg.db.BeginRw(ctx)
+		if err != nil {
+			return fmt.Errorf("Failed to open tx, %w", err)
+		}
+		defer tx.Rollback()
+	}
+
+	// batch until which witness has been generated and saved to db
+	witnessProgress, err := stages.GetStageProgress(tx, stages.Witness)
+	if err != nil {
+		return fmt.Errorf("save stage progress error: %v", err)
+	}
+
+	// batch which has been processed by stage_batches
+	latestBlock, err := stages.GetStageProgress(tx, stages.Execution)
+	if err != nil {
+		return fmt.Errorf("could not retrieve batch no progress")
+	}
+
+	hermezDbReader := hermez_db.NewHermezDbReader(tx)
+
+	batchesProgress, err := hermezDbReader.GetBatchNoByL2Block(latestBlock)
+	if err != nil {
+		return fmt.Errorf("could not retrieve batch for latest block")
+	}
+
+	if witnessProgress >= batchesProgress {
+		log.Info("All batch witness already in db, skipping")
+		return nil
+	}
+
+	// witness generation should start from stored witness of batch in db + 1
+	witnessFromBatch := witnessProgress + 1
+	// witness generation should happen till batch of latest block -1
+	// as there might be more blocks incoming for this batch
+	witnessToBatch := batchesProgress - 1
+
+	wg := NewWitnessGenerator(cfg.tmpDir, cfg.historyV3, cfg.agg, cfg.blockReader, cfg.chainConfig, cfg.engine)
+
+	for b := witnessFromBatch; b <= witnessToBatch; b++ {
+		highestBlockInBatch, err := hermezDbReader.GetHighestBlockInBatch(b)
+		if err != nil {
+			return err
+		}
+
+		if highestBlockInBatch == 0 {
+			continue
+		}
+
+		nextStageProgress, err := stages.GetStageProgress(tx, stages.HashState)
+		if err != nil {
+			return err
+		}
+		if nextStageProgress < highestBlockInBatch {
+			// Skip this stage
+			return nil
+		}
+
+		log.Info(fmt.Sprintf("[%s] Starting witness generation", logPrefix), "currentBatch", b, "highestBatch", witnessToBatch)
+		w, err := wg.GenerateWitnessByBatch(tx, ctx, b, false)
+		if err != nil {
+			return err
+		}
+
+		hermezDb := hermez_db.NewHermezDb(tx)
+		err = hermezDb.WriteWitnessByBatchNo(b, w)
+		if err != nil {
+			return err
+		}
+
+		err = stages.SaveStageProgress(tx, stages.Witness, b)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
```

Comment:
We probably need to remove old witnesses from db, e.g. keep only last 1000 (configurable through cli) witnesses, in order to prevent the db from growing too big.